packages/react-components/src/templates/TagsPageBody.tsx

```diff
@@ -29,8 +41,44 @@ const MessageBody: React.FC<{ title: string; body: string }> = ({
   </main>
 );
 
+export enum TagFieldByEntity {
+  'research-output' = 'keywords',
+  user = 'expertiseAndResourceTags',
+  event = 'calendar',
+}
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+const assert = <T extends never>() => undefined;
+type TypeEqualityGuard<A, B> = Exclude<A, B> | Exclude<B, A>;
+
+const EntityCard: React.FC<TagsPageBodyProps['results'][number]> = ({
+  ...data
+}): JSX.Element => {
+  assert<
+    TypeEqualityGuard<keyof typeof TagFieldByEntity, CRNTagSearchEntities>
+  >;
+
+  if (TagFieldByEntity['research-output'] in data) {
+    return <SharedResearchCard {...data} />;
+  }
+
+  if (TagFieldByEntity.user in data) {
+    return <PeopleCard {...data} />;
+  }
+
+  if (TagFieldByEntity.event in data) {
+    return <EventCard {...eventMapper(data)} />;
+  }
+
+  return <ErrorCard />;
+};
+
 interface TagsPageBodyProps {
-  readonly results: ListResearchOutputResponse['items'];
+  readonly results: (
```

Comment:
Could we use the type that we exported from the model here?

Reply:
committed here [review: use TagSearchResponse from model in TagsPageBody](https://github.com/yldio/asap-hub/pull/3856/commits/5991aaf54f35559f84380018178e877d07d7be3f)